lib/ansible/parsing/dataloader.py

```diff
@@ -233,35 +236,33 @@ def path_dwim(self, given):
         given = unquote(given)
         given = to_text(given, errors='surrogate_or_strict')
 
-        if given.startswith(u"/"):
-            return os.path.abspath(given)
-        elif given.startswith(u"~"):
-            return os.path.abspath(os.path.expanduser(given))
+        if RE_ABS_PATH.search(given):
+            path = unfrackpath(given)
         else:
             basedir = to_text(self._basedir, errors='surrogate_or_strict')
-            return os.path.abspath(os.path.join(basedir, given))
+            path = unfrackpath(os.path.join(basedir, given))
+
+        return path
 
     def _is_role(self, path):
-        ''' imperfect role detection, roles are still valid w/o main.yml/yaml/etc '''
+        ''' imperfect role detection, roles are still valid w/o tasks|meta/main.yml|yaml|etc '''
 
-        isit = False
         b_path = to_bytes(path, errors='surrogate_or_strict')
         b_upath = to_bytes(unfrackpath(path), errors='surrogate_or_strict')
 
-        for suffix in (b'.yml', b'.yaml', b''):
-            b_main = b'main%s' % (suffix)
-            b_tasked = b'tasks/%s' % (b_main)
-
-            if (
-                b_path.endswith(b'tasks') and
-                os.path.exists(os.path.join(b_path, b_main)) or
-                os.path.exists(os.path.join(b_upath, b_tasked)) or
-                os.path.exists(os.path.join(os.path.dirname(b_path), b_tasked))
-            ):
-                isit = True
-                break
+        for finddir in (b'meta', b'tasks'):
+            for suffix in (b'.yml', b'.yaml', b''):
+                b_main = b'main%s' % (suffix)
+                b_tasked = b'%s/%s' % (finddir, b_main)
 
-        return isit
+                if (
+                    RE_TASKS.search(path) and
```

Comment:
I think the same comment about using substring matching instead of a regex applies here as it did earlier.  however, the regex looks different from the substring match it's replacing.  Can you tell me why that is and what it's supposed to be matching now?